libevm/JitVM.cpp

```diff
@@ -16,129 +16,83 @@ static_assert(sizeof(Address) == sizeof(evm_uint160be),
 {
-	return *reinterpret_cast<evm_uint256be*>(&_h);
+	return *reinterpret_cast<evm_uint256be const*>(&_h);
 }
 
-inline u256 asUint(evm_uint256be _n)
+inline u256 fromEvmC(evm_uint256be const& _n)
 {
 	return fromBigEndian<u256>(_n.bytes);
 }
 
-inline h256 asHash(evm_uint256be _n)
-{
-	return h256(&_n.bytes[0], h256::ConstructFromPointer);
-}
-
-void evm_query(
+void queryState(
 	evm_variant* o_result,
 	evm_env* _opaqueEnv,
 	evm_query_key _key,
-	evm_variant const* _arg
+	evm_uint160be const* _addr,
+	evm_uint256be const* _storageKey
 ) noexcept
 {
 	auto &env = *reinterpret_cast<ExtVMFace*>(_opaqueEnv);
+	Address addr = fromEvmC(*_addr);
 	switch (_key)
 	{
-	case EVM_ADDRESS:
-		o_result->address = toEvmC(env.myAddress);
-		break;
-	case EVM_CALLER:
-		o_result->address = toEvmC(env.caller);
-		break;
-	case EVM_ORIGIN:
-		o_result->address = toEvmC(env.origin);
-		break;
-	case EVM_GAS_PRICE:
-		o_result->uint256be = toEvmC(env.gasPrice);
-		break;
-	case EVM_COINBASE:
-		o_result->address = toEvmC(env.envInfo().author());
-		break;
-	case EVM_DIFFICULTY:
-		o_result->uint256be = toEvmC(env.envInfo().difficulty());
-		break;
-	case EVM_GAS_LIMIT:
-		// TODO: Handle overflow / exception
-		o_result->int64 = static_cast<int64_t>(env.envInfo().gasLimit());
-		break;
-	case EVM_NUMBER:
-		// TODO: Handle overflow / exception
-		o_result->int64 = static_cast<int64_t>(env.envInfo().number());
-		break;
-	case EVM_TIMESTAMP:
-		// TODO: Handle overflow / exception
-		o_result->int64 = static_cast<int64_t>(env.envInfo().timestamp());
-		break;
 	case EVM_CODE_BY_ADDRESS:
 	{
-		auto addr = fromEvmC(_arg->address);
 		auto &code = env.codeAt(addr);
 		o_result->data = code.data();
 		o_result->data_size = code.size();
 		break;
 	}
 	case EVM_CODE_SIZE:
-	{
-		auto addr = fromEvmC(_arg->address);
 		o_result->int64 = env.codeSizeAt(addr);
 		break;
-	}
 	case EVM_BALANCE:
-	{
-		auto addr = fromEvmC(_arg->address);
 		o_result->uint256be = toEvmC(env.balance(addr));
 		break;
-	}
-	case EVM_BLOCKHASH:
-		o_result->uint256be = toEvmC(env.blockHash(_arg->int64));
-		break;
 	case EVM_SLOAD:
 	{
-		auto key = asUint(_arg->uint256be);
-		o_result->uint256be = toEvmC(env.store(key));
+		auto storageKey = fromEvmC(*_storageKey);
+		o_result->uint256be = toEvmC(env.store(storageKey));
 		break;
 	}
 	case EVM_ACCOUNT_EXISTS:
-	{
-		auto addr = fromEvmC(_arg->address);
 		o_result->int64 = env.exists(addr);
 		break;
 	}
-	case EVM_CALL_DEPTH:
-		o_result->int64 = env.depth;
-		break;
-	}
 }
 
-void evm_update(
+void updateState(
 	evm_env* _opaqueEnv,
 	evm_update_key _key,
+	evm_uint160be const* _addr,
 	evm_variant const* _arg1,
 	evm_variant const* _arg2
 ) noexcept
 {
+	(void) _addr;
 	auto &env = *reinterpret_cast<ExtVMFace*>(_opaqueEnv);
+	assert(fromEvmC(*_addr) == env.myAddress);
 	switch (_key)
 	{
 	case EVM_SSTORE:
 	{
-		auto index = asUint(_arg1->uint256be);
-		auto value = asUint(_arg2->uint256be);
+		auto index = fromEvmC(_arg1->uint256be);
```

Comment:
`auto` & `fromEvmC` combination is a little confusing here, it's not clear to what we convert